torch/distributed/algorithms/_comm_hooks/default_hooks.py

```diff
@@ -83,43 +83,83 @@ def allreduce_hook(state: DefaultState, grad: torch.Tensor):
         state (DefaultState): State information, configures pre- and post-division factors
         grad (torch.Tensor): A gradient for the local batch that needs to be communicated across ranks.
     """
+    # Average grad by pre-division factor. Together pre- and post-division factors
+    # lead to an overall averaging by world_size, required for consistency with PyTorch DDP.
+    # This is a two-step process to avoid potential underflow and overflow.
     if state.gradient_predivide_factor > 1:
         grad.div_(state.gradient_predivide_factor)
     dist.all_reduce(grad, group=state.process_group)
+    # Average grad by post-division factor. Together pre- and post-division factors
+    # lead to an overall averaging by world_size, required for consistency with PyTorch DDP.
+    # This is a two-step process to avoid potential underflow and overflow.
     if state.gradient_postdivide_factor > 1:
         grad.div_(state.gradient_postdivide_factor)
 
-def lower_precision_hook(prec: torch.dtype, state: LowPrecisionState, grad: torch.Tensor):
-    grad.data = grad.data.to(prec)
-    allreduce_hook(state, grad)
-    _decompress(state, grad)
+def reduce_scatter_hook(state: DefaultState, grad: torch.Tensor, output: torch.Tensor):
+    r"""
+    This FSDP communication hook implements ``reduce_scatter`` algorithm for
+    sharded FSDP strategies and a necessary pre- and post-division of gradients.
+
+    Args:
+        state (DefaultState): State information, configures pre- and post-division factors
+        grad (torch.Tensor): A full gradient for the local batch that needs to be
+        communicated across ranks.
+        output (torch.Tensor): Stores a single shard of the gradient after ``reduce_scatter``.
+    """
+    # Average grad by pre-division factor. Together pre- and post-division factors
+    # lead to an overall averaging by world_size, required for consistency with PyTorch DDP.
+    # This is a two-step process to avoid potential underflow and overflow.
+    if state.gradient_predivide_factor > 1:
+        grad.div_(state.gradient_predivide_factor)
+    dist._reduce_scatter_base(
+        output, grad, group=state.process_group
+    )
+    # Average grad's shard by post-division factor. Together pre- and post-division factors
+    # lead to an overall averaging by world_size, required for consistency with PyTorch DDP.
+    # This is a two-step process to avoid potential underflow and overflow.
+    if state.gradient_postdivide_factor > 1:
+        output.div_(state.gradient_postdivide_factor)
 
-def fp16_compress_hook(state: LowPrecisionState, grad: torch.Tensor):
+def lower_precision_hook(prec: torch.dtype, state: LowPrecisionState, grad: torch.Tensor, output: torch.Tensor = None):
```

Comment:
Shall we just call this `low_precision_hook` for naming consistency?

Reply:
Also, if this is just a subroutine for the actual hooks like `fp16_compress_hook`, then we should make it private (preceding underscore). If every method that calls it expects to actually pass a value for `output`, then we should not make `output` defaulted to `None`; rather, we let the methods that call this method to have `output` defaulted to `None`.

Reply:
@awgu for `NO_SHARD` case `output` is not expected and is not passed, so I've decided to default it to `None` here.

Reply:
What I mean is that `output` is already defaulted to `None` for the public methods, so there is no need to make it a default argument for `lower_precision_hook()`.

Reply:
@awgu I see, thanks for clarification